New extension: module

gabrielschulhof opened the pull request:
This extension provides the following facilities:
  - registering modules,
  - registering module resolvers, and
  - an API for retrieving a module instance given its name.

A module is defined as a global static structure containing a pointer
to a string which is the name of the module, and a pointer to a function
which will be called when an instance of the module is needed.

A module resolver is a function that accepts a string holding the name
of the module and returns a `jerry_value_t` if the module was found, or
zero if it was not. If it returns a non-zero value with the error flag
set then the API will pass it through without caching.

This extension provides a built-in module resolver which attempts to
load modules that follow the above module definition.

The API provided by this extension invokes all module resolvers in
sequence to attempt to resolve the name of a single module. After one
resolver returns a non-zero `jerry_value_t`, the API stops iterating
over the remaining resolvers and caches the value if its error flag is
not set. It then returns the `jerry_value_t`. The API will return a
`jerry_value_t` containing an error indicating that the module was not
found if it reaches the end of the list of resolvers. The error it
returns has an extra property `"moduleName"` the value of which is a
string containing the name of the module that the API was asked to
resolve.

JerryScript-DCO-1.0-Signed-off-by: Gabriel Schulhof gabriel.schulhof@intel.com

glistening commented:
@gabrielschulhof This PR's module syntax looks similar to nodejs's `require`. Do you have reason to use this syntax? (e.g., backward compatibility to existing nodejs style code?) Are you going to implement node.js compliant module system? Why don't we follow ECMA2015 specification module syntax? JerryScript is JavaScript Engine, so at least for me, it seems natural to make jerryscript ECMA specification compliant.

gabrielschulhof commented:
@glistening I'm not providing any new language structures in this PR. I'm simply providing a way of converting a string to a `jerry_value_t` with module semantics. It is indeed similar to Node.js if that's how you choose to expose it to JS - which is what my unit test does.

I'm certainly in favour of implementing ES6 modules in JerryScript. I can imagine JerryScript exposing an API such as
```c
typedef jerry_value_t (*jerry_module_handler_t) (const jerry_char_t *);
```
and
```c
void jerry_set_module_handler (jerry_module_handler_t handler);
```
or even
```c
void jerry_init_with_module_handler (jerry_init_flag_t flag, jerry_module_handler_t handler);
```
which can be satisfied by `jerryx_module_resolve ()` as provided by this PR.

I am not sufficiently familiar with the JerryScript interpreter to write the parser for the ES6 syntax bits concerning modules, however, I'm fairly certain that we can use this PR to implement ES6 modules if/when the parser is ready and uses an external hook as described above.

martijnthe commented:
@glistening:

> Why don't we follow ECMA2015 specification module syntax? JerryScript is JavaScript Engine, so at least for me, it seems natural to make jerryscript ECMA specification compliant.

This was discussed briefly in the community call of 2017/04/19.
Here is a snippet from the meeting notes – See https://groups.io/g/jerryscript-dev/message/7

> "... Was discussing with Gabriel, working on building a utility on top of this API, like a module manager, similar to require(). Requires bookkeeping. ES6 spec has abstract module record (for import/export). Modules have state, need to have a handle related to that state. Native bindings also need this. Would it make sense to add something to jerry-core that implements abstract module record? So we can add import/require()/native bindings on top of it."

I remember @zherczeg or someone else from the Szeged folks saying something along the lines of: let's put it into `jerryx` first, we can later move it into `jerry-core`.

My hope is that by putting the basic foundation of module management into JerryScript, that it will become easier to mix and match native binding code from different projects.

ES2015 modules, `import`, etc is a lot more work to build out, so I'm considering this just a start of it.

gabrielschulhof commented:
@glistening for example, we could write a module resolver for text files (JS source code):
```c
JERRYX_MODULE_RESOLVER(load_source_code)
{
  jerry_value_t ret = 0;
  FILE *module_file = NULL;
  const char absolute_path[4096] = MODULES_ROOT_DIRECTORY;
  strncpy (&absolute_path[sizeof (MODULES_ROOT_DIRECTORY)],
    name, 4096 - sizeof (MODULES_ROOT_DIRECTORY));
  module_file = fopen(absolute_path, "r");
  if (!module_file)
  {
    return 0;
  }
  /* load and ret = jerry_eval () the file */
  fclose (module_file);
  return ret;
}
```
Now, if the text file is an ES6 module, then `ret` will be the result of having loaded an ES6 module, and if the module has imports, then the `load_source_code ()` module resolver will be called recursively, or iteratively from a main loop, or whatever.

glistening commented:
@gabrielschulhof, @martijnthe  Oh, thank you for the history and explanatory examples.

gabrielschulhof commented:
@martijnthe we needn't necessarily move the module resolution into jerry-core. We need only that jerry-core expose a hook for informing that some JS code has requested a module.

gabrielschulhof commented:
Oh, OK, so in my example above we probably can't use `jerry_eval ()` because JS' `eval ()` is [not supposed to be able to handle modules](http://2ality.com/2014/09/es6-modules-final.html#eval-and-modules). So, we need a new Jerry API like `jerry_eval_module (const jerry_char_t *module_definition)` which understands kerywords like `import` and `export` and such.

The bottom line is that I think interpreting new keywords and integrating the concept of modules into the JS engine is something that needs to be done in core, whereas the mechanisms for resolving modules, whether native or text, can be "hooked out" to extension(s).

zherczeg commented on docs/11.EXT-REFERENCE-MODULE.md:
````diff
@@ -0,0 +1,116 @@
+
+```c
+jerry_value_t
+jerryx_module_resolve (const jerry_char_t *name);
````
What about `jerry_x_get_module` ?

gabrielschulhof replied:
The convention for extension seems to be `jerryx_<extension>_what_the_api_does ()`. Also, `resolve` is used in the spec to refer to loading modules.

zherczeg commented on jerry-ext/include/jerryscript-ext/section.impl.h:
```diff
@@ -0,0 +1,84 @@
+ * @param name the name of the section without quotes
+ */
+#ifdef __MACH__
+#define JERRYX_SECTION_NAME(name) "__DATA," #name
```
What about those systems where this option is not available?

gabrielschulhof replied:
You mean for systems that do not support linker sections?

An alternative for such systems is to use a constructor, for which we now have support in jerry-libc. However, constructors are also stored in a special linker section, so that too might not work either.

An uglier alternative if all else fails is to use code generation and parse the source code to extract `JERRYX_MODULE()` declarations.

Either way, we can define `JERRYX_MODULE()` in various different platform-specific ways.

zherczeg replied:
Ok, so without this macro the whole module system does not work? No workaround at all?

gabrielschulhof replied:
@zherczeg I guess a workaround is possible, but it's hard to test, because I have to build jerry-ext with `-U__GNUC__` (cf. https://github.com/jerryscript-project/jerryscript/issues/1890#issuecomment-311889136)

zherczeg replied:
Hm, come to think of it it would be good overall to test the whole project without GNUC macro. A travis job could be added.

zherczeg replied:
(Btw I am happy with having no native modules if certain compiler features are missing. The loading from file system is still useful.)

gabrielschulhof replied:
@akosthekiss @galpeter @dbatyai @zherczeg @LaszloLango @glistening @martijnthe @jiangzidong @bzsolt @robertsipka @janjongboom So, can we land this without the workaround and add the workaround as part of the solution to https://github.com/jerryscript-project/jerryscript/issues/1890?

gabrielschulhof replied:
@zherczeg I tried precisely that (adding a Travis job) using the command line I mention in my comment, but I'm not familiar enough with the engine to fix those warnings.

gabrielschulhof replied:
@zherczeg Also, `-U__GNUC__` != extensions disabled, because if `__attribute__((something))` appears unconditionally, gcc will still accept that. So, it remains our responsibility to provide for the case where `__GNUC__` is not defined and then to test for that case using travis.

I actually wonder if Ubuntu has a C compiler that does not provide GNU extensions. Then we could do a travis job where we use that as the compiler instead of the default one.

martijnthe replied:
I'm fine not having the workaround here yet.

zherczeg commented on jerry-ext/include/jerryscript-ext/section.impl.h:
```diff
@@ -0,0 +1,84 @@
+ * @code
+ * int index;
+ * widget_t *current_widget_p;
+ * for JERRYX_SECTION_ITERATE (widget_array, index, current_widget_p)
```
For me this is a bit ugly. We should invent a more common macro.

zherczeg commented on jerry-ext/module/module.c:
```diff
@@ -0,0 +1,215 @@
+  return ret;
+} /* jerryx_module_create_error */
+
+
```
Single newline after functions.

zherczeg commented on jerry-ext/module/module.c:
```diff
@@ -0,0 +1,215 @@
+                            jerry_value_t module_name, /**< key at which to cache the module */
+                            jerry_value_t module) /**< the module to cache */
+{
+  jerry_value_t ret = jerry_set_property (cache, module_name, module);
```
Do you use a jerry object for caching module names? It is a bit inefficient if many modules are loaded.

gabrielschulhof replied:
Using a jerry object avoids having to use `jmem_heap_alloc_block ()`. I can certainly revert to using `jmem_heap_alloc_block ()` but what data structure should I use, especially if we're considering that there may be many modules? I was using a linked list before, but that would probably be even more inefficient.

I'm certainly happy to implement, like a prefix tree or something.

zherczeg replied:
Hm, this is an implementation detail ,so we can change it anytime later. We could start using objects.

zherczeg commented on jerry-ext/module/module.c:
```diff
@@ -0,0 +1,215 @@
+  }
+
+  /* Try each resolver in the section until one manages to return non-zero. */
+  for JERRYX_SECTION_ITERATE (jxm_resolvers, index, resolver_p)
```
What about dynamic modules (i.e. comes from a JS file).

gabrielschulhof replied:
That can be implemented separately: https://github.com/jerryscript-project/jerryscript/pull/1863#issuecomment-304797823

zherczeg replied:
We still need a plan from API perspective.

gabrielschulhof replied:
That's true, but, aside from support for ES6 modules syntax (like `import` and `export`) I don't think we need anything new from core. So, we just need to agree to throw our support behind a convention for loading modules. We should consider supporting Node.js and AMD.

For example, if we do what Node.js does, then we wrap the contents of a .js file in
```JS
(function (exports, require, module, __filename, __dirname) {
/* contents of JS file */
});
```
And then
```c
jerry_value_t the_function = jerry_eval (wrapped_file_contents);
jerry_call_function (the_function, ...);
```
and retrieve the exports.

AMD uses a different approach. For AMD we need not wrap the original JS in anything, but we need to provide the `define ()` method.
```JS
define( /* optional string with the module name */, [ "dep1", "dep2", "dep3" ],
  function( dep1, dep2, dep3 ) {
    /* Module body goes here. dep1, dep2, and dep3 are loaded asynchronously,
     * and the return value is the module. */
  } );
```

There's an effort to unify all of these into a single universal module definition: https://github.com/umdjs/umd

This is all a bit hairy, to say the least.

martijnthe commented:
@janjongboom could you / other mbedjs people take a look at this and provide feedback? It's the foundation for a native module system. If everyone starts using it, it will hopefully make it easier to re-use native module code in different projects.

jiangzidong commented on jerry-ext/include/jerryscript-ext/section.impl.h:
```diff
@@ -0,0 +1,54 @@
+ * @param name the name of the section (without quotes)
+ * @param type the type of the elements stored in the array
+ *
+ * This macro needs to be placed before any occurrence of JERRYX_SECTION_ITERATE, for a given section.
```
`This macro needs to be placed before any iterations inside a given section`

jiangzidong commented on jerry-ext/module/module.c:
```diff
@@ -0,0 +1,218 @@
+  jerry_value_t context = jerry_create_object ();
+  *((jerry_value_t *) user_data_p) = jerry_acquire_value (context);
+  jerry_release_value (context);
+} /* jerryx_module_manager_init */
```
why not just
```
{
*((jerry_value_t *) user_data_p) = jerry_create_object ();
}
```

gabrielschulhof replied:
I wasn't sure actually. I thought `jerry_acquire_value ()` was needed to keep the value around indefinitely.

jiangzidong replied:
I think they are equivalent.
object created by `jerry_create_object ` will never be GCed unless you call the `jerry_release_value`.

jiangzidong commented on jerry-ext/module/module.c:
```diff
@@ -0,0 +1,218 @@
+static jerry_value_t
+jerryx_module_check_cache (jerry_value_t cache, /**< cache from which to attempt to retrieve the module by name */
+                          jerry_value_t module_name) /* JerryScript string value holding the module name */
+{
```
What about using `jerry_get_property (cache, module_name) `, if ret is not undefined, then the module is in tha cache, otherwise not.

gabrielschulhof replied:
Actually, a module could intentionally evaluate to `undefined`. The result of loading a module could be anything, including `undefined`. That is, the entire gamut of possible JS values must be available to modules. We cannot pick a possible value and reserve it for ourselves.

jiangzidong replied:
Yes, that's why I suggest return error if there is no proper resolver.  Your bool way should also work.

BTW, why do we need multiple resolvers?

gabrielschulhof replied:
For multiple module formats. In fact, my chaining idea from below probably won't work, because we need to assemble the chain at some point.

jiangzidong commented on jerry-ext/module/module.c:
```diff
@@ -0,0 +1,218 @@
+    {
+      /* If we find the module by its name we load it and cache it if it has an on_resolve () and complain otherwise. */
+      ret = (module_p->on_resolve) ? module_p->on_resolve () :
+      jerryx_module_create_error (JERRY_ERROR_TYPE, on_resolve_absent, name);
```
The indention is confusing,

gabrielschulhof replied:
Sorry! If I add two spaces before `jerryx_module_create_error` then check-vera.sh complains.

jiangzidong replied:
what about this way
```
a = (b ? c
       : d);
```
which is the example of `08.CODING-STANDARDS.md`

gabrielschulhof replied:
It complains about that too. The best it lets me do is
```
a = (b ? c
: d);
```

jiangzidong replied:
@zherczeg Do you know why?

jiangzidong commented on jerry-ext/module/module.c:
```diff
@@ -0,0 +1,218 @@
+  for (index = 0, resolver_p = &__start_jxm_resolvers[0];
+  &__start_jxm_resolvers[index] < __stop_jxm_resolvers;
+  index++, resolver_p = &__start_jxm_resolvers[index])
+  {
```
Why do we need multiple resolver-functions? And How can user make sure the order those functions?

I think one resolver function is enough, and user can set its own one if they like.

gabrielschulhof replied:
The idea is that you can have multiple module systems. For example, one resolver takes care of native modules, whereas another takes care of loading text files.

I guess we could chain them if we do something like
```c
typedef jerry_value_t (jerryx_module_resolver_t) (const jerry_char_t *);
jerryx_module_resolver_t
jerryx_module_set_resolver (jerryx_module_resolver_t new_resolver);
```
where `jerryx_module_set_resolver ()` returns the previous resolver - kinda like signal handlers.
The order must not matter.

jiangzidong replied:
Though you said `order must not matter`, but people may provides resolves whose older matters, and it can work well. But when other rebuild the same project, the order may change and resulting in strange error.

Sorry that I don't totally understand the reason why we need multiple resolver in one project. Could you give me an example?

gabrielschulhof replied:
One example is the built-in loader that uses the list of modules from the section `jxm_modules`. Another example is the one I gave @glistening [above](https://github.com/jerryscript-project/jerryscript/pull/1863#issuecomment-304797823).

As for evaluation order, the idea is that no two resolvers resolve the same module. If somebody adds module "xyzzy.js" to their project, then adds another module "xyzzy.c" which uses `JERRYX_MODULE()` internally, then it is undefined which one they're going to get when they attempt to `jerryx_module_resolve ("xyzzy")`.

martijnthe replied:
https://github.com/jerryscript-project/jerryscript/pull/1863#issuecomment-304797823 is IMO a very good reason to want to have multiple resolvers.

gabrielschulhof commented on docs/11.EXT-REFERENCE-MODULE.md:
```diff
@@ -0,0 +1,116 @@
+Define a module resolver. An invocation of this macro is to be followed by the body of a module resolver. The macro
+declares a function of the name given in `cb_name` which accepts a single parameter `const jerry_char_t *name`. It is
+expected that the function returns a non-zero value of type `jerry_value_t` if the resolver is able to associate `name`
+with such a value, or zero if it is not. If the function returns non-zero `jerryx_module_resolve ()` will cache the
```
0 indicates that that the resolver was unable to resolve the module and that the next resolver should be consulted.

gabrielschulhof replied:
Nevertheless, I can change the signature to return `bool` and the result in an out parameter.

jiangzidong replied:
"return bool or error obj" is better to me

janjongboom commented:
@gabrielschulhof @martijnthe Will look at it in the next few days. Traveling right now. But I applaud the idea!

jiangzidong commented:
@gabrielschulhof Can I say that all modules which use JERRYX_MODULE to register must be handled by native_resolver, right?

gabrielschulhof commented:
@jiangzidong correct.

LaszloLango commented on jerry-ext/include/jerryscript-ext/module.h:
```diff
@@ -0,0 +1,55 @@
+#define JERRYX_MODULE(module_name, on_resolve_cb)                        \
+  static jerryx_module_t _module JERRYX_SECTION_ATTRIBUTE(jxm_modules) = \
+  {                                                                      \
+    .name = ((jerry_char_t *)#module_name),                              \
```
Missing space: `((jerry_char_t *) #module_name)`

LaszloLango commented on jerry-ext/include/jerryscript-ext/module.h:
```diff
@@ -0,0 +1,55 @@
+#define JERRYX_MODULE_RESOLVER(cb_name)                                              \
+  static bool cb_name (const jerry_char_t *name, jerry_value_t *result);             \
+  static bool (*__static_global_ ## cb_name) (const jerry_char_t *, jerry_value_t *) \
+  JERRYX_SECTION_ATTRIBUTE(jxm_resolvers) = cb_name;                                 \
```
ditto: `JERRYX_SECTION_ATTRIBUTE (jxm_resolvers)`

zherczeg commented on docs/11.EXT-REFERENCE-MODULE.md:
```diff
@@ -0,0 +1,117 @@
+- return value - `jerry_value_t` representing the module that was loaded, or the error that occurred in the process.
+
+
+# Modules data types
```
Module data types

zherczeg commented on jerry-ext/include/jerryscript-ext/module.h:
```diff
@@ -0,0 +1,55 @@
+ * Define a module. The section name is cryptic ("jxm") because it has to be less than 16 bytes on Mach-O.
+ */
+#define JERRYX_MODULE(module_name, on_resolve_cb)                        \
+  static jerryx_module_t _module JERRYX_SECTION_ATTRIBUTE(jxm_modules) = \
```
Could this be const data?

zherczeg commented on jerry-ext/module/module.c:
```diff
@@ -0,0 +1,215 @@
+  /* Look for the module by its name in the list of module definitions. */
+  for (index = 0, module_p = &__start_jxm_modules[0];
+  &__start_jxm_modules[index] < __stop_jxm_modules;
+  index++, module_p = &__start_jxm_modules[index])
```
This code does not follow the style.

gabrielschulhof replied:
Skimming over the coding standards doc I don't see any section that applies to long for-statement headers. How should I indent this? Keep in mind that `check-vera.sh` does not now complain about indentation.

gabrielschulhof replied:
NM. Found it. Let's see if check-vera.sh allows it.

zherczeg commented on jerry-ext/module/module.c:
```diff
@@ -0,0 +1,215 @@
+/**
+ * Declare and define the default module resolver - one which examines what modules are defined in the above linker
+ * section and loads one that matches the requested name, caching the result for subsequent requests using the context
+ * data mechanism.
```
Does this work if no static modules are defined?

gabrielschulhof replied:
Good question. I'll add a test.

gabrielschulhof replied:
[Added](https://github.com/jerryscript-project/jerryscript/pull/1863/files#diff-56a70d8121afefa3ff42d36017116ef3)

zherczeg commented on jerry-ext/include/jerryscript-ext/module.h:
```diff
@@ -0,0 +1,55 @@
+  static bool cb_name (const jerry_char_t *name, jerry_value_t *result);             \
+  static bool (*__static_global_ ## cb_name) (const jerry_char_t *, jerry_value_t *) \
+  JERRYX_SECTION_ATTRIBUTE (jxm_resolvers) = cb_name;                                 \
+  static bool cb_name (const jerry_char_t *name, jerry_value_t *result)
```
What is the priority order of the module resolvers? In case if multiple resolvers reacts to the same name.

gabrielschulhof replied:
There is no priority order. The first one that resolves the module wins. If multiple resolvers provide the same module, it is undefined which one you're going to get. It's up to project maintainers to keep their list of modules clean.

I've added a `*Note*:` into the doc.

gabrielschulhof replied:
For example, if somebody wants to load module "http", and two differently structured modules resolved by different resolvers both provide "http", then you're gonna get "http" one way or the other. If you're confused as to why the JS API provided by the resulting module is totally different from what you expected, well, check your build scripts and don't load two of the same thing into your project.

gabrielschulhof replied:
Note that in npm this is not a problem, since each package has a unique name, and there are already two different package structures: native addon packages, and plain JS packages.

zherczeg replied:
For a generic module selector, defining the order of resolvers is very important. We cannot expect that names are unique, and we need to let people to define resolve order for symbol clashes. E.g. a module on a file system has a higher priority than a builtin module, or vice versa. We need to figure out something for this.

martijnthe replied:
Hmm, yeah I tend to agree that I'd control over the order in which the resolvers are called.
I'm mostly in interested in having one resolver for native modules and one for source file modules. But I don't want an app's JS code to be able to "override" native module with a source file module. Therefore, the native module resolver should be run before the source file module resolver. For my use case it's fine to decide the priority at build-time, I don't have a need to resolve clashes at runtime or anything like that.

Maybe add a priority argument to `JERRYX_MODULE_RESOLVER`, so one can write something like this?

```
#ifndef MY_RESOLVER_PRIORITY
#  define MY_RESOLVER_PRIORITY (1)
#endif

JERRYX_MODULE_RESOLVER(my_resolver, MY_RESOLVER_PRIORITY)
{
...
}
```

gabrielschulhof replied:
I guess the resolve order is a bit like a search path for modules. For example, first we search the linker section, then we search the file system.

We could give resolvers names, and accept a comma-separated list of names. We could then re-arrange the `"jxm_resolvers"` linker section so that, the next time around, presuming that the requested order will be the same, the resolvers will be in the correct order.

So, like,
```c
jerryx_module_resolve("name", "built-in,js-fs-loader");
```
`NULL` would mean whatever order is there. If there are more resolvers than there are items in the list, and the listed ones are unable to resolve the module, the rest would be consulted in whatever order they're left in the section after the rearrangement completes.

This would make the resolver kinda like a meta-module.

martijnthe replied:
Passing the ordered list of resolvers to use to the `jerryx_module_resolve` would work for my purposes. I'm not a huge fan of passing the resolvers as a comma separated list of strings though..
Maybe it makes sense make the resolver a visible/usable symbol, as opposed to being static, and just it directly in the `jerryx_module_resolve` call?

gabrielschulhof replied:
That could become unwieldy:
```c
static const jerryx_module_resolver_t resolvers = {
#ifdef HAVE_JS_FS_RESOLVER
  jerryx_module_resolver_js_fs,
#endif /* HAVE_JS_FS_RESOLVER */
#fidef HAVE_SOME_OTHER_RESOLVER
  jerryx_module_some_other_resolver,
#endif /* HAVE_SOME_OTHER_RESOLVER */
};
```
and
```c
jerry_value_t = jerryx_module_resolve("something",
                                      resolvers,
                                      sizeof(resolvers)/sizeof(resolvers[0]));
```
Though I highly doubt that the array of resolvers will go too-too large.

gabrielschulhof replied:
TBH I think passing `NULL` as the second parameter would work 99.9% of the time, because any given module is provided by exactly one of the resolvers.

gabrielschulhof replied:
A completely different alternative would be not to expose the resolver at all, and to maintain its capabilities as part of this extension ourselves.

martijnthe replied:
> That could become unwieldy

Doesn't look too bad to me...

Maybe it makes sense to have 2 separate APIs, `jerryx_module_resolve` (simple, no control over ordering) and `jerryx_module_resolve_using_resolvers` (complex, control over ordering) or something.

gabrielschulhof replied:
I'll go with the array and the default `NULL` if it's not that unwieldy. I'll have to change `JERRYX_MODULE_RESOLVER` to accept an existing function rather than declaring a new function.

martijnthe replied:
cool, thanks

zherczeg replied:
I see the reason behind a random order for built-in modules (stored in a linker section for simplicity), but the number of resolvers should be low (1-3 maximum), so I would prefer a static list of them. This list (and its length) should be passed when the resolvers are initialized. The general resolver simply goes through this list.

gabrielschulhof replied:
OK, there's another wrinkle: The results of module resolution are cached. So, if you pass the same `name` to `jerryx_module_resolve ()` at a later time, you're going to get whatever module was loaded the first time around, even if you specify a different order of resolvers, which would result in a different module getting cached if this were the first time around.

gabrielschulhof replied:
@zherczeg 
How do you mean "when the resolvers are initialized"? We do not currently have a separate initialization step for the resolvers. With my current (not yet pushed) modification, the list of resolvers is passed to `jerryx_module_resolve ()`, so it becomes
```c
jerry_value_t jerryx_module_resolve (const jerry_char_t *name,
                                     const jerryx_module_resolver_t *resolvers,
                                     size_t count);
```

zherczeg replied:
Sounds good!

zherczeg replied:
I don't see any problem with cached resolvers. If we ever find a module, we need to use it forever (in its current state).

bzsolt commented:
@gabrielschulhof could you please refer to the new docfile in the [tools/update-webpage.sh](https://github.com/jerryscript-project/jerryscript/blob/master/tools/update-webpage.sh)? (this script generates the proper files for the webpage)
Just add the following two lines (or something similar):

```
EXT_REFERENCE_MODULE_MD="11.EXT-REFERENCE-MODULE.md"
...
titles[$EXT_REFERENCE_MODULE_MD]="'Extension API: Module Handling'"
```

gabrielschulhof commented:
@bzsolt Done.

LaszloLango commented on docs/11.EXT-REFERENCE-MODULE.md:
```diff
@@ -0,0 +1,108 @@
+jerry_value_t
+jerryx_module_resolve (const jerry_char_t *name, /**< name of the module to load */
+                       jerryx_module_resolver_t *resolvers_p, /**< list of resolvers */
+					   size_t resolver_count); /**< number of resolvers in @p resolvers */
```
wrong indentation

LaszloLango commented on jerry-ext/include/jerryscript-ext/module.h:
```diff
@@ -0,0 +1,56 @@
+  {                                                                      \
+    .name = ((jerry_char_t *) #module_name),                              \
+    .on_resolve = (on_resolve_cb)                                        \
+  };
```
Please align the backslash characters to the same column

gabrielschulhof commented:
@zherczeg I switched to a list of resolvers explicitly passed. I got rid of the `"jxm_resolvers"` section, so the list cannot be empty.

LaszloLango commented on jerry-ext/module/module.c:
```diff
@@ -0,0 +1,213 @@
+ * data mechanism.
+ */
+bool jerryx_module_native_resolver (const jerry_char_t *name, /**< name of the module */
+                                    jerry_value_t *result) /**< out - where to put the resulting module instance */
```
`jerry_value_t *result) /**< [out] where to put the resulting module instance */`

LaszloLango commented on jerry-ext/module/module.c:
```diff
@@ -0,0 +1,213 @@
+    {
+      /* If we find the module by its name we load it and cache it if it has an on_resolve () and complain otherwise. */
+      (*result) = ((module_p->on_resolve) ? module_p->on_resolve ()
+      : jerryx_module_create_error (JERRY_ERROR_TYPE, on_resolve_absent, name));
```
```c
(*result) = ((module_p->on_resolve) ? module_p->on_resolve ()
                                    : jerryx_module_create_error (JERRY_ERROR_TYPE, on_resolve_absent, name));
```

jiangzidong commented on docs/11.EXT-REFERENCE-MODULE.md:
```diff
@@ -0,0 +1,108 @@
+#  Module API
+
```
Could you also introduce the concept of "module" and "module reducer" here

gabrielschulhof replied:
@jiangzidong please see below.

jiangzidong commented on docs/11.EXT-REFERENCE-MODULE.md:
```diff
@@ -0,0 +1,108 @@
+**Summary**
+
+Load a copy of a module into the current context or return one that was already loaded if it is found.
+
```
We should also tell the user the meaning  of the order in resolvers_list

gabrielschulhof replied:
@jiangzidong please see below.

jiangzidong commented on docs/11.EXT-REFERENCE-MODULE.md:
````diff
@@ -0,0 +1,108 @@
+                                    jerry_value_t *result) /**< out - where to put the resulting module instance */
+```
+- `name` - the name of the module to find
+- `result` - out - place where to store the resulting JS value
````
`place where to store the resulting module instance`

jiangzidong commented on docs/11.EXT-REFERENCE-MODULE.md:
```diff
@@ -0,0 +1,108 @@
+
+**Summary**
+
+Helper macro to define a module. Declares a variable of type `jerryx_module_t` as global static and places it into the
```
`Helper macro to define a JerryScript module`

jiangzidong commented on docs/11.EXT-REFERENCE-MODULE.md:
```diff
@@ -0,0 +1,108 @@
+**Summary**
+
+Structure to define a module
+
```
`module` -> `JerryScript module`

jiangzidong commented on docs/11.EXT-REFERENCE-MODULE.md:
```diff
@@ -0,0 +1,108 @@
+
+**Summary**
+
+The resolver for JerryScript modules. A pointer to this function can be passed in the second parameter to `jerryx_module_resolve` to search for the module among the JerryScript modules built into the binary.
```
`The resolver for JerryScript modules` , we should put it clearer. 
1) what is jerryscript module? I think it is **ONLY** the module which is declared by `JERRYX_MODULE`, right? (We can also put the description in the beginning of the doc)
2) For other types of module, user should provide their own resolver

gabrielschulhof replied:
@jiangzidong the introduction at the top of the file now covers both these topics.

jiangzidong commented on jerry-ext/module/module.c:
```diff
@@ -0,0 +1,213 @@
+
+/**
+ * Attempt to retrieve a module by name from a cache, and return zero if not found.
+ */
```
return false

jiangzidong commented on tests/unit-ext/module-empty.c:
```diff
@@ -0,0 +1,70 @@
+/* Copyright JS Foundation and other contributors, http://js.foundation
```
what about renaming `module-empty` => "test-ext-module-empty", so that all the unittest case under /unit-ext have similar name pattern

bzsolt commented on tools/update-webpage.sh:
```diff
@@ -32,6 +32,7 @@ DEBUGGER_MD="07.DEBUGGER.md"
 CODING_STANDARDS_MD="08.CODING-STANDARDS.md"
 EXT_REFERENCE_ARG_MD="09.EXT-REFERENCE-ARG.md"
 EXT_REFERENCE_HANDLER_MD="10.EXT-REFERENCE-HANDLER.md"
+EXT_REFERENCE_MODULE_MD="11.EXT-REFERENCE-MODULE.mpd"
```
Thank you for wiring the doc, but there is a typo in the extension of filename: `mpd` -> `md`.

bzsolt commented on tools/update-webpage.sh:
```diff
@@ -45,6 +46,7 @@ titles[$DEBUGGER_MD]="Debugger"
 titles[$CODING_STANDARDS_MD]="Coding Standards"
 titles[$EXT_REFERENCE_ARG_MD]="'Extension API: Argument Validation'"
 titles[$EXT_REFERENCE_HANDLER_MD]="'Extension API: External Function Handlers'"
+titles[$EXT_REFERENCE_MODULE_MD]="'Extension API: Module support'"
```
Please start the word "support" with capital letter like the others.

gabrielschulhof commented:
@jiangzidong Thanks!

LaszloLango commented on docs/11.EXT-REFERENCE-MODULE.md:
````diff
@@ -0,0 +1,178 @@
+
+```c
+jerry_value_t
+jerryx_module_resolve (const jerry_char_t *name, /**< name of the module to load */
````
Please remove the comments. You described the parameters below.

LaszloLango commented on docs/11.EXT-REFERENCE-MODULE.md:
````diff
@@ -0,0 +1,178 @@
+**Prototype**
+
+```c
+bool jerryx_module_native_resolver (const jerry_char_t *name, /**< name of the module */
````
ditto

zherczeg commented on docs/11.EXT-REFERENCE-MODULE.md:
```diff
@@ -0,0 +1,178 @@
+      (*result) = jerry_eval (js_file_contents, file_size, false);
+
+      /* We release the memory holding the contents of the file. */
+	  free (js_file_contents);
```
Misaligned line.

zherczeg commented on docs/11.EXT-REFERENCE-MODULE.md:
````diff
@@ -0,0 +1,178 @@
+
+We can now load JavaScript files:
+```c
+jerryx_module_resolver_t resolvers =
````
const static

zherczeg commented on docs/11.EXT-REFERENCE-MODULE.md:
```diff
@@ -0,0 +1,178 @@
+
+# Module helper macros
+
+## JERRYX_MODULE
```
These are called native modules above.

zherczeg commented on jerry-ext/include/jerryscript-ext/module.h:
```diff
@@ -0,0 +1,56 @@
+typedef bool (*jerryx_module_resolver_t) (const jerry_char_t *name, jerry_value_t *result);
+
+/**
+ * The implementation of the resolver for modules as defined above
```
Full stop after sentence.

zherczeg commented on jerry-ext/include/jerryscript-ext/module.h:
```diff
@@ -0,0 +1,56 @@
+  };
+
+/**
+ * Type of a module resolver function
```
Ditto.

zherczeg commented on jerry-ext/include/jerryscript-ext/module.h:
```diff
@@ -0,0 +1,56 @@
+} jerryx_module_t;
+
+/**
+ * Define a module. The section name is cryptic ("jxm") because it has to be less than 16 bytes on Mach-O.
```
What is Mach-O? 3 character is pretty short, we have at least 13 characters more for a bit more descriptive name/

akosthekiss replied:
Mach-O is Mach object file format

LaszloLango commented:
Please rebase to the current master

gabrielschulhof commented:
@LaszloLango I'm working on a way of assembling the array of module definitions by providing a tool that recognizes the pattern `JERRYX_MODULE (...)` and outputting a C file that creates a global variable initialized to the list of modules. I will rebase when I have that in place.

gabrielschulhof commented:
@LaszloLango this is for the case where `__attribute__((section(...)))` is unavailable - IOW `__GNUC__` is not defined.

gabrielschulhof commented:
@zherczeg I have now addressed your review comments. The module system will not work `#ifndef __GNUC__`, although I cannot currently test that because of https://github.com/jerryscript-project/jerryscript/issues/1890#issuecomment-311889136. @LaszloLango I have rebased.

zherczeg commented on docs/12.EXT-REFERENCE-MODULE.md:
```diff
@@ -0,0 +1,182 @@
+#  Module API
+
+**Note**: This extension depends on compiler support because it uses `__attribute__` extensions provided only by GCC and
+LLVM/clang. For other compilers, no support has been added yet.
```
Please update this to the current status.

gabrielschulhof replied:
How do you mean? gcc/clang-only *is* the current status.

zherczeg replied:
Please describe what happens when other compilers are used.

gabrielschulhof replied:
Actually, I just realized that only the JerryScript modules will be affected. If you create modules using some other mechanism and provide a resolver, that will work. So, I'll document *that*.

zherczeg commented on docs/12.EXT-REFERENCE-MODULE.md:
````diff
@@ -0,0 +1,182 @@
+```c
+jerry_value_t
+jerryx_module_resolve (const jerry_char_t *name,
+                       *resolvers_p,
````
Type of resolvers?

zherczeg commented on docs/12.EXT-REFERENCE-MODULE.md:
```diff
@@ -0,0 +1,182 @@
+{
+  jerry_char_t *name; /**< name of the module */
+  jerry_value_t (*on_resolve) (void); /**< function that returns a new instance of the module */
+} jerryx_module_t;
```
I think this structure is only used by native modules. Furthermore do we need to make it public? I mean it is needed to be public, but people should never use its members directly. We should emphasize this.

gabrielschulhof replied:
I guess we need not document this structure, but only the macro. We can then leave a comment in the header emphasizing that the structure not be used directly.

zherczeg replied:
Ok.

zherczeg commented on jerry-ext/include/jerryscript-ext/module.h:
```diff
@@ -0,0 +1,60 @@
+{
+  jerry_char_t *name; /**< name of the module */
+  jerry_value_t (*on_resolve) (void); /**< function that returns a new instance of the module */
+} jerryx_module_t;
```
The native module only part should should be hidden in some way when the support is missing. And than type name should be `jerryx_native_module_t`

gabrielschulhof replied:
I'm not sure we should call it `jerryx_native_module_t` - after all, it's a "JerryScript module" and that's how we document it.

zherczeg replied:
Is it used for non-native modules?

zherczeg commented on jerry-ext/include/jerryscript-ext/module.h:
```diff
@@ -0,0 +1,60 @@
+ * specially-named linker section "jerryx_modules" where the JerryScript module resolver
+ * jerryx_module_native_resolver () will look for it.
+ */
+#define JERRYX_MODULE(module_name, on_resolve_cb)                                 \
```
JERRYX_NATIVE_MODULE

gabrielschulhof commented:
It shouldn't be used for non-native modules, because those will have their
own definition mechanism and will have their own resolver too.

On Thu, Jun 29, 2017 at 12:23 PM, Zoltan Herczeg <notifications@github.com>
wrote:

> *@zherczeg* commented on this pull request.
> ------------------------------
>
> In jerry-ext/include/jerryscript-ext/module.h
> <https://github.com/jerryscript-project/jerryscript/pull/1863#discussion_r124751575>
> :
>
> > + */
> +
> +#ifndef JERRYX_MODULE_H
> +#define JERRYX_MODULE_H
> +
> +#include "jerryscript.h"
> +#include "jerryscript-ext/section.impl.h"
> +
> +/**
> + * Declare the structure used to define a module.
> + */
> +typedef struct
> +{
> +  jerry_char_t *name; /**< name of the module */
> +  jerry_value_t (*on_resolve) (void); /**< function that returns a new instance of the module */
> +} jerryx_module_t;
>
> Is it used for non-native modules?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/jerryscript-project/jerryscript/pull/1863#discussion_r124751575>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA7k0SrgOkUk_7_VXjPKSWDB4qMc1_-Eks5sI20dgaJpZM4Npg2f>
> .
>

zherczeg commented on docs/12.EXT-REFERENCE-MODULE.md:
```diff
@@ -0,0 +1,177 @@
+**Note**: The `JERRYX_MODULE` macro depends on compiler support because it uses `__attribute__` extensions provided
+only by GCC and LLVM/clang to ensure that JerryScript's own module resolver (`jerryx_module_native_resolver`) will
+find a module that was declared with `JERRYX_MODULE`. If such compiler support is absent, currently
+`jerryx_module_native_resolver` will fail to find a module declared with `JERRYX_MODULE`.
```
To tell you the truth I don't really like the compiles but does not work at runtime approach. That is awful to debug. I would prefer a compiler error to raise the attention. What I would prefer: conditional definition of macros / structs / functions which related to native modules, and all of these types should have `native` in their name to make their role clear.

gabrielschulhof replied:
Ha! That's exactly what I had at first :) I just put `#ifdef __GNUC__` at the top of each file involved. OK. I'll add the `native_` and switch back to that approach.

zherczeg replied:
I would alao add a define for this feature when it is enabled, so people can test it at compile time and do a workaround if needed:

    #ifdef __GNUC__
    ...
    #define JERRYX_NATIVE_MODULES_SUPPORTED
    ...
    #endif

gabrielschulhof commented:
@zherczeg Please have another look. I have defined `JERRYX_NATIVE_MODULES_SUPPORTED` and I have substantially updated the documentation to reflect that native modules are only available when that macro is defined while also emphasizing that the module resolver continues to work even if native modules are unavailable, as long as you provide your own resolver.

zherczeg commented on jerry-ext/include/jerryscript-ext/module.h:
```diff
@@ -0,0 +1,72 @@
+/**
+ * Declare the signature for the module initialization function.
+ */
+typedef jerry_value_t (*jerryx_module_on_resolve_t) (void);
```
Shall we add native prefix to this type?

gabrielschulhof replied:
✓

zherczeg commented on jerry-ext/include/jerryscript-ext/section.impl.h:
```diff
@@ -0,0 +1,56 @@
+ */
+
+#ifndef JERRYX_SECTION_IMPL_H
+#define JERRYX_SECTION_IMPL_H
```
I would add a JERRYX_NATIVE_MODULES_SUPPORTED guard here, or where this file is included.

gabrielschulhof replied:
`jerryscript-ext/module.h` refers to this file. I have moved that reference to after the `#ifdef JERRYX_NATIVE_MODULES_SUPPORTED`.

gabrielschulhof replied:
https://github.com/jerryscript-project/jerryscript/pull/1863/files#diff-9ae8ab785adef517e883b35c916b144cR26

zherczeg commented on jerry-ext/include/jerryscript-ext/section.impl.h:
```diff
@@ -0,0 +1,56 @@
+#else /* !__MACH__ */
+#define JERRYX_SECTION_DECLARE(name, type)                    \
+  extern const type __start_ ## name[] __attribute__((weak)); \
+  extern const type __stop_ ## name[] __attribute__((weak));
```
This is just a question: why these symbols are weak? Do we expect them to be overwritten by strong symbols?

gabrielschulhof replied:
I believe it's so that if there is no such section, then their value is `NULL` and iterating over the array will short-circuit.

zherczeg commented on jerry-ext/module/module.c:
```diff
@@ -0,0 +1,216 @@
+#include "jerryscript-ext/module.h"
+
+static const jerry_char_t *module_name_property_name = (jerry_char_t *) "moduleName";
+static const jerry_char_t *on_resolve_absent = (jerry_char_t *) "Module on_resolve () must not be NULL";
```
Am I correct that this error text is used when native modules are used? We should not define unused symbols.

gabrielschulhof replied:
I have moved this into the `#ifdef JERRYX_NATIVE_MODULES_SUPPORTED` block.

zherczeg commented on jerry-ext/module/module.c:
```diff
@@ -0,0 +1,216 @@
+    {
+      /* If we find the module by its name we load it and cache it if it has an on_resolve () and complain otherwise. */
+      (*result) = ((module_p->on_resolve) ? module_p->on_resolve ()
+                                          : jerryx_module_create_error (JERRY_ERROR_TYPE, on_resolve_absent, name));
```
Do we complain about NULL onresolve, but not about NULL names? It is a bit inconsistent to me. Hm, I am fine to ignore modules without names (does strcmp supports NULL arguments?).

gabrielschulhof replied:
Good point! I don't think a `NULL` name is OK, because it cannot be requested. I mean, sure, in JS you can write `var nullModule = require(null);` but c'mon, seriously, I can't imagine that being a feature somebody would want.

gabrielschulhof replied:
OTOH, I don't think the resolver should throw an error if it encounters a module with a `NULL` name because that would poison the rest of the linker section, preventing *any* native modules following the defective module from being loaded. So, I will simply add a `NULL`-check before the `strcmp()` because I don't believe it's `NULL`-safe.

LaszloLango commented on docs/12.EXT-REFERENCE-MODULE.md:
````diff
@@ -0,0 +1,187 @@
+```c
+/* An example resolver for JavaScript files */
+bool
+load_and_evaluate_js_file(const jerry_char_t *name, jerry_value_t *result)
````
missing space before `(`

LaszloLango commented on docs/12.EXT-REFERENCE-MODULE.md:
````diff
@@ -0,0 +1,187 @@
+
+**Example**
+```c
+/* An example resolver for JavaScript files */
````
I think this comment is unnecessary here.

LaszloLango commented on docs/12.EXT-REFERENCE-MODULE.md:
```diff
@@ -0,0 +1,187 @@
+    js_file_contents = malloc (file_size);
+    if (js_file_contents)
+    {
+      /* We read the file into memory and call jerry_eval (), assigning the result to the out-parameter. */
```
Not sure why, but GitHub marks the end of the line with a red color for me. Is there a trailing whitespace at the end of the line?

gabrielschulhof replied:
Nope. I wager it's some kind of parse error.

LaszloLango commented on jerry-ext/include/jerryscript-ext/module.h:
```diff
@@ -0,0 +1,72 @@
+ */
+#define JERRYX_NATIVE_MODULE(module_name, on_resolve_cb)                                 \
+  static const jerryx_native_module_t _module JERRYX_SECTION_ATTRIBUTE(jerryx_modules) = \
+  {                                                                               \
```
Please put the backslash characters to the same column.

gabrielschulhof replied:
✓

gabrielschulhof commented:
@zherczeg @LaszloLango I have addressed your review comments.

zherczeg commented on jerry-ext/include/jerryscript-ext/module.h:
```diff
@@ -0,0 +1,72 @@
+
+#ifdef JERRYX_NATIVE_MODULES_SUPPORTED
+#include "jerryscript-ext/section.impl.h"
+/**
```
Newline before comment.

zherczeg commented on jerry-ext/include/jerryscript-ext/module.h:
```diff
@@ -0,0 +1,72 @@
+ * Declare the JerryScript module resolver so that it may be added to an array of jerryx_module_resolver_t items and
+ * thus passed to jerryx_module_resolve.
+ */
+bool jerryx_module_native_resolver (const jerry_char_t *name, jerry_value_t *result);
```
Newline after function declaration.

gabrielschulhof replied:
Added

zherczeg commented on jerry-ext/include/jerryscript-ext/section.impl.h:
```diff
@@ -0,0 +1,56 @@
+/* Copyright JS Foundation and other contributors, http://js.foundation
```
The name section.impl.h does not refer to modules. It is clear that It is an internal header and should not be used, but it is still part of the public API.

gabrielschulhof replied:
I didn't intend for it to be related to modules. It contains macros that help with linker sections, whatever purpose those linker sections may serve.

LaszloLango commented:
LGTM
